src/main/java/bio/terra/service/snapshot/flight/setpublic/SetSnapshotPublicStep.java

```
@@ -0,0 +1,59 @@
+package bio.terra.service.snapshot.flight.setpublic;
+
+import bio.terra.common.exception.ForbiddenException;
+import bio.terra.common.exception.NotFoundException;
+import bio.terra.common.iam.AuthenticatedUserRequest;
+import bio.terra.service.auth.iam.IamResourceType;
+import bio.terra.service.auth.iam.IamRole;
+import bio.terra.service.auth.iam.IamService;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import bio.terra.stairway.exception.RetryException;
+import java.util.UUID;
+
+public class SetSnapshotPublicStep implements Step {
+  private final UUID snapshotId;
+  private final boolean setPublic;
+  private final AuthenticatedUserRequest userReq;
+  private final IamService iamService;
+
+  public SetSnapshotPublicStep(
+      UUID snapshotId, boolean setPublic, AuthenticatedUserRequest userReq, IamService iamService) {
+    this.snapshotId = snapshotId;
+    this.setPublic = setPublic;
+    this.userReq = userReq;
+    this.iamService = iamService;
+  }
+
+  private StepResult setPublic(boolean publicVal) {
+    try {
+      iamService.setPolicyPublicV2(
+          userReq.getToken(),
+          IamResourceType.DATASNAPSHOT,
+          snapshotId,
+          IamRole.READER.name(),
+          publicVal);
+      return StepResult.getStepResultSuccess();
+    } catch (ForbiddenException | NotFoundException e) {
+      return new StepResult(
+          StepStatus.STEP_RESULT_FAILURE_FATAL,
+          new ForbiddenException(
+              "User is not authorized to set this resource as public, contact Terra support for assistance.",
```

we can return this exception message for a 404 because 404 is returned if the resource does not exist or there is not permission on it, and if the resource does not exist the get public call in the controller will fail, meaning at this point we know it exists but there is not permission on it.